guides/hack/11-built-in-types/36-enum-class-label.md

```diff
@@ -0,0 +1,102 @@
+safe to call if all the values of the enumeration are distinct.
+
+Enum classes provides a way to do this by using the newly introduced *Enum Class Label* expressions. For each value defined in an enum class, a corresponding
+label is defined. A label is a handle to access the related value. Think of it as an indirect access. Consider the following example:
+
+```EnumClassLabel.definition.hack no-auto-output
+
+// We are using int here for readability but it works for any type
+enum class E : int {
+  int A = 42;
+  int B = 42;
+}
+```
+
+This example defines two constants `E::A: \HH\MemberOf<E, int>` and `E::B: \HH\MemberOf<E, int>`. Enum class labels add more definitions to the mix:
+
+- `E#A: \HH\EnumClass\Label<E, int>` is the label to access `E::A`
+- `E#B: \HH\EnumClass\Label<E, int>` is the label to access `E::B`
+- `E::nameOf` is a static method expecting a label and returning its string representation: `E::nameOf(E#A) = "A"`
+- `E::valueOf` is a static method expecting a label and returning its value: `E::valueOf(E#A) = E::A`
+
+So we can rewrite the earlier example in a more resilient way:
+```EnumClassLabel.example.hack no-auto-output
+<<file:__EnableUnstableFeatures('enum_class_label')>> // temp
+
+function full_print(\HH\EnumClass\Label<E, int> $label) : void {
+  echo E::nameOf($label) . " ";
+  echo E::valueOf($label) . "\n";
+}
+
+function partial_print(\HH\MemberOf<E, int> $value) : void {
+  echo $value . "\n";
+}
+```
+Now, `full_print(E#A)` will echo `A 42` and `full_print(E#B)` will echo `B 42`.
+
+## Full v. Short Labels
+
+We refer to labels like `E#A` as *fully qualified* labels: the programmer wrote the full enum class name. However there are some situations where Hack can infer
+this part, and we can save the programmer that bit. For example, the previous calls could be written as `full_print(#A)` and `full_print(#B)`, leaving `E` implicit.
+This is only allowed when there is enough type information to infer the right enum class name. For example, `$x = #A` is not allowed and will result in a type error.
+
+### Special case of function calls
+
+When the first argument of a function is a label, we provide an alternative notation to call it. This was done to reflect some generated code patterns this feature helped removed:
+```EnumClassLabel.alt.hack no-auto-output
+<<file:__EnableUnstableFeatures('enum_class_label')>> // temp
+
+function set<T>(\HH\EnumClass\Label<E, T> $label, T $data) : void {
+  // setting $data into some storage using $label as a key
+}
+
+// all these calls are equivalent
+function all_the_same(): void {
+  set(E#A, 42);
+  set(#A, 42);
+  set#A(42);
+}
+```
+
+As you can see, the short name can be written *before* the opening parenthesis of the function call. This only works for a single label argument, which must be the first one.
+
+## Known corner cases
+
+### The `#` character is no longer a one line comment
```

Comment:
Minor, but we use the terminology "single-line comment"

Reply:
huh, if this true, we should remove  the '#' in "Single-line comments start with // or #, and end with a newline." in 
 [hack/source-code-fundamentals/comments](https://docs.hhvm.com/hack/source-code-fundamentals/comments)

Reply:
Indeed, that change has been approved a couple month ago now and `#` is no longer a single-line comment.